spatial/r3/affine.go

```diff
@@ -0,0 +1,316 @@
+	m.x13 = (a.x02*a.x13*a.x20 - a.x03*a.x12*a.x20 + a.x03*a.x10*x22 - x00*a.x13*x22 - a.x02*a.x10*a.x23 + x00*a.x12*a.x23) * d
+	m.x20 = (x11*a.x23*a.x30 - a.x13*a.x21*a.x30 + a.x13*a.x20*a.x31 - a.x10*a.x23*a.x31 - x11*a.x20*x33 + a.x10*a.x21*x33) * d
+	m.x21 = (a.x03*a.x21*a.x30 - a.x01*a.x23*a.x30 - a.x03*a.x20*a.x31 + x00*a.x23*a.x31 + a.x01*a.x20*x33 - x00*a.x21*x33) * d
+	m.d22 = (a.x01*a.x13*a.x30-a.x03*x11*a.x30+a.x03*a.x10*a.x31-x00*a.x13*a.x31-a.x01*a.x10*x33+x00*x11*x33)*d - 1
+	m.x23 = (a.x03*x11*a.x20 - a.x01*a.x13*a.x20 - a.x03*a.x10*a.x21 + x00*a.x13*a.x21 + a.x01*a.x10*a.x23 - x00*x11*a.x23) * d
+	m.x30 = (a.x12*a.x21*a.x30 - x11*x22*a.x30 - a.x12*a.x20*a.x31 + a.x10*x22*a.x31 + x11*a.x20*a.x32 - a.x10*a.x21*a.x32) * d
+	m.x31 = (a.x01*x22*a.x30 - a.x02*a.x21*a.x30 + a.x02*a.x20*a.x31 - x00*x22*a.x31 - a.x01*a.x20*a.x32 + x00*a.x21*a.x32) * d
+	m.x32 = (a.x02*x11*a.x30 - a.x01*a.x12*a.x30 - a.x02*a.x10*a.x31 + x00*a.x12*a.x31 + a.x01*a.x10*a.x32 - x00*x11*a.x32) * d
+	m.d33 = (a.x01*a.x12*a.x20-a.x02*x11*a.x20+a.x02*a.x10*a.x21-x00*a.x12*a.x21-a.x01*a.x10*x22+x00*x11*x22)*d - 1
+	return m
+}
+
+func makeAffineRotation(r Rotation) Affine {
+	if r == (Rotation{}) {
+		return Affine{}
+	}
+	w, i, j, k := r.Real, r.Imag, r.Jmag, r.Kmag
+	ii := 2 * i * i
+	jj := 2 * j * j
+	kk := 2 * k * k
+	wi := 2 * w * i
+	wj := 2 * w * j
+	wk := 2 * w * k
+	ij := 2 * i * j
+	jk := 2 * j * k
+	ki := 2 * k * i
+	return Affine{
+		d00: -jj - kk, x01: ij - wk, x02: ki + wj,
+		x10: ij + wk, d11: -ii - kk, x12: jk - wi,
+		x20: ki - wj, x21: jk + wi, d22: -ii - jj,
+	}
+}
+
+// transpose returns the transposed transform.
+func (a Affine) transpose() Affine {
+	return Affine{
+		d00: a.d00, x01: a.x10, x02: a.x20, x03: a.x30,
+		x10: a.x01, d11: a.d11, x12: a.x21, x13: a.x31,
+		x20: a.x02, x21: a.x12, d22: a.d22, x23: a.x32,
+		x30: a.x03, x31: a.x13, x32: a.x23, d33: a.d33,
+	}
+}
+
+// sliceCopy returns a copy of the transform's data
+// in row major storage format. It returns 16 elements.
+func (a Affine) sliceCopy() []float64 {
+	return []float64{
+		a.d00 + 1, a.x01, a.x02, a.x03,
+		a.x10, a.d11 + 1, a.x12, a.x13,
+		a.x20, a.x21, a.d22 + 1, a.x23,
+		a.x30, a.x31, a.x32, a.d33 + 1,
+	}
+}
+
+// isIdentity returns true if receiver is the identity transform.
+func (a Affine) isIdentity() bool {
+	// The zero value of Affine is guaranteed to be the identity.
+	return a == Affine{}
+}
+
+// IsZero returns true if a is the zero transform.
+func (a Affine) IsZero() bool {
+	return a == zeroAffine
+}
+
+// Warp is an affine transformation that scales and shears (skews) the domain.
+//
+// The zero value of Warp corresponds to the identity transform, thus the diagonal
+// elements of Warp are stored subtracted 1. This means that a transformation
+// that scales all components of a vector by 3 is constructed as follows:
+//  w := Warp{XX:2, YY:2, ZZ:2}
+type Warp struct {
+	XX, XY, XZ float64
+	YX, YY, YZ float64
+	ZX, ZY, ZZ float64
+}
```

Comment:
I'm not a fan of this. It leaves all the work up to the user and is an end-run around the work that's gone in to not exporting the elements of `Affine`. I agree that the complexity of representing skew is high, but I would prefer that this be how it's represented. Probably `type Skew struct { XY, XZ, YX, YZ, ZX, ZY float64 }`.

Reply:
> It leaves all the work up to the user

I'm not sure I understand- what work do you mean? 

> end-run around the work that's gone in to not exporting the elements of Affine.

The Affine's top non-zero 3x3 elements may or may not encode Warp data. They are functionally different.

> `type Skew struct { XY, XZ, YX, YZ, ZX, ZY float64 }`

This identifier is already taken by the skew symmetric matrix constructor.

Reply:
> I'm not sure I understand- what work do you mean?

There is work that needs to be done by the user to make sure they don't shoot themselves in the foot; this is confusing:

"This means that a transformation that scales all components of a vector by 3 is constructed as follows: `w := Warp{XX:2, YY:2, ZZ:2}`"

The reason is clear when you know the code, but it will be the cause of issues.

It would be better to not have both scaling and skew together in the API. A `Mat` with the note that the diagonal is ignored would work, 

> This identifier is already taken by the skew symmetric matrix constructor.

This is unfortunate. What was the reason for introducing the skew-symmetric function? (it wasn't in the proposal that provoked the PR).

I'm starting to worry about the complexity of the API that is growing here.

Reply:
> Mat with the note that the diagonal is ignored would work,

OK, this can work.

> This is unfortunate. What was the reason for introducing the skew-symmetric function? (it wasn't in the proposal that provoked the PR).

See https://github.com/gonum/gonum/issues/1754. Skew was in the proposal. In hindsight it should have been a method - this would have simplified package-level API and also prevented allocations on every Skew call. The reason I included it as a function was that the `Skew` operation to me was a basic vector commodity, much like `Cross`.

> I'm starting to worry about the complexity of the API that is growing here.

It might be a little early to set `ComposeAffine` in stone- I'm still learning a lot about Affine transforms. We can leave it for a future PR.